### ISSUE TITLE
Ghostrunner crashes during texture injection

### ISSUE THREAD

**Issue by sl-einert** [SpecialKO/SpecialK#12]
Hi,
when launching and dumping textures of Ghostrunner everything works fine.
But when injecting the original textures, even without modifications, the game crashes when loading these textures.

![grafik](https://user-images.githubusercontent.com/59532465/180058582-a4ddd4b6-85eb-4a16-8b5c-1259169c4ef0.png)

Sadly the logs do not give much more info:
`07/20/2022 18:29:49.081: D:\Steam Games\steamapps\common\Ghostrunner\Ghostrunner\Binaries\Win64\EOSSDK-Win64-Shipping.dll:  Shutdown handler: initalize.
07/20/2022 18:30:02.041: D:\Steam Games\steamapps\common\Ghostrunner\Ghostrunner\Binaries\Win64\Ghostrunner-Win64-Shipping.exe:  LowLevelFatalError [File:Unknown] [Line: 258] 
Direct3DDevice->CreateShaderResourceView(TextureResource,&SRVDesc,ShaderResourceView.GetInitReference()) failed 
 at E:/UnrealEngineOML_Source/Engine/Source/Runtime/Windows/D3D11RHI/Private/D3D11Texture.cpp:1315 
 with error E_INVALIDARG
07/20/2022 18:30:26.099: D:\Steam Games\steamapps\common\Ghostrunner\Ghostrunner\Binaries\Win64\Ghostrunner-Win64-Shipping.exe:  LowLevelFatalError [File:Unknown] [Line: 907] 
Rendering thread exception:

LowLevelFatalError [File:Unknown] [Line: 258] 
Direct3DDevice->CreateShaderResourceView(TextureResource,&SRVDesc,ShaderResourceView.GetInitReference()) failed 
 at E:/UnrealEngineOML_Source/Engine/Source/Runtime/Windows/D3D11RHI/Private/D3D11Texture.cpp:1315 
 with error E_INVALIDARG
`
I am using the pretty much the default settings for SpecialK with activated injection as for other games where these settings work fine.

Am I missing some settings that need to be set or is this game just not compatible with SpecialK Injections?

Best regards

**Comment by sl-einert** [comment]
I just back checked with another game which made troubles in the past and it is the same error, but maybe easier to reproduce as it was free in the Epic Store.
https://www.pcgamingwiki.com/wiki/Mothergunship

**Comment by Kaldaien** [comment]
Unreal Engine loads many of its textures using a `_TYPELESS` format initially, and in order to dump them to disk a type gets assigned. When you go to re-inject them, the typed format isn't castable and this sort of thing tends to happen.

For example:

![image](https://user-images.githubusercontent.com/13346629/180436340-696abeff-79a9-4a7a-a89f-beb0b1ba4555.png)

I have added a number of workarounds for this in the past, but I don't think they're adequate. Can you confirm that the textures you have dumped and tried to re-inject actually have a type (i.e. `_UNORM`)?

**Comment by sl-einert** [comment]
You are right. The textures format in the metadata end with `_TYPELESS.` I checked the files in Visual Studio and they all end with `_UNORM`

**Comment by sl-einert** [comment]
I did some experiments, maybe that helps you:
I was able to inject textures with the correct format. I demonstrate it here with `99E85597.dds` of MotherGunShip which can be found in the Mission Room on the ground.
![grafik](https://user-images.githubusercontent.com/59532465/181043077-f9047466-5879-4313-95b9-3a74601a1471.png)
```
Dumped Name:    ***\My Mods\SpecialK\Profiles\TowerOfGuns2-Win64-Shipping.exe\SK_Res/dump/textures/TowerOfGuns2-Win64-Shipping.exe/\Compressed_99E85597.dds.dds
Texture:        99e85597
Dimensions:     (512x256)
Format:         076 (DXGI_FORMAT_BC3_TYPELESS)
MipLODs:        10
Compressed:     Yes
sRGB:           Yes
----------------
BindFlags:      0x0008
Usage:          0x0000
CPUAccessFlags: 0x00
Misc:           0x00
ArraySize:      01
```

Visual Studio shows us the current format as `BC3_UNORM`.
![grafik](https://user-images.githubusercontent.com/59532465/181040834-4d076c7c-4754-41c5-8dd0-37b1820be9ea.png)

Because I encountered strange color changes with Visual Studio(but for testing you can use it), I switched to Paint.net and use it to convert it to `BC3_UNORM_SRGB`:
![grafik](https://user-images.githubusercontent.com/59532465/181041525-9fcf1138-7094-40e7-a731-d99258b21388.png)
![grafik](https://user-images.githubusercontent.com/59532465/181041709-b5f7f5e2-314a-4819-b2d5-98428f19ff1a.png)

For verification we see now in VS that the Compression is now `BC3_UNORM_SRGB`:
![grafik](https://user-images.githubusercontent.com/59532465/181042026-75771d66-34e9-468a-b815-ce4f7f6a90e2.png)

Which now is perfectly injected:
![grafik](https://user-images.githubusercontent.com/59532465/181042537-13f2a610-31ce-4f99-9657-8917087959b2.png)


Interesting is that even textures that do not show the sRBG Attribute in the txt-File require the `_UNORM_SRGB` Format:
```
Dumped Name:    ***\My Mods\SpecialK\Profiles\TowerOfGuns2-Win64-Shipping.exe\SK_Res\dump\textures\TowerOfGuns2-Win64-Shipping.exe\Compressed_AFAD5FDC.dds
Texture:        afad5fdc::8bf290a6
Dimensions:     (256x256)
Format:         070 (DXGI_FORMAT_BC1_TYPELESS)
MipLODs:        09
----------------
BindFlags:      0x0008
Usage:          0x0000
CPUAccessFlags: 0x00
Misc:           0x00
ArraySize:      01
```
![grafik](https://user-images.githubusercontent.com/59532465/181048617-ccc1d7e9-819e-4247-a628-76ca729e9861.png)


Maybe SpecialK could dump those textures as `_UNORM_SRGB`, but I don't know if that is possible or another solution makes more sense. 

Hope that helps you :-)
